common/src/worldgen/horosphere.rs

```diff
@@ -0,0 +1,319 @@
+    fn is_horosphere_pos_valid(
+        graph: &Graph,
+        node_id: NodeId,
+        horosphere_pos: &MVector<f32>,
+    ) -> bool {
+        // See `should_propagate` for an explanation of what the mip between a horosphere position and
+        // a plane's normal signifies.
+        Side::iter().all(|s| s.normal().mip(&horosphere_pos) < 1.0)
+            && (graph.descenders(node_id)).all(|(s, _)| s.normal().mip(horosphere_pos) < -1.0)
+    }
+
+    /// Returns a vector representing a uniformly random horosphere within a certain distance to the origin.
+    /// This distance is set up to ensure that `is_horosphere_pos_valid` would always return false if it were any futher,
+    /// ensuring that this function does not artificially restrict which horospheres can be created. Rejection
+    /// sampling is used to more precisely finetune the list of allowed horospheres.
+    fn random_horosphere_pos(rng: &mut Pcg64Mcg) -> MVector<f32> {
+        // Pick a w-coordinate whose probability density function is `p(w) = w`. By trial and error,
+        // this seems to produce horospheres with a uniform and isotropic distribution.
+        // TODO: Find a rigorous explanation for this. We would want to show that the probability density is unchanged
+        // when an isometry is applied.
+        let w = sqrtf(rng.random::<f32>()) * Self::MAX_OWNED_HOROSPHERE_W;
+
+        // Uniformly pick spherical coordinates from a unit sphere
+        let cos_phi = rng.random::<f32>() * 2.0 - 1.0;
+        let sin_phi = sqrtf(1.0 - cos_phi * cos_phi);
+        let theta = rng.random::<f32>() * std::f32::consts::TAU;
+
+        // Construct the resulting vector.
+        MVector::new(
+            w * sin_phi * cosf(theta),
+            w * sin_phi * sinf(theta),
+            w * cos_phi,
+            w,
+        )
+    }
+
+    /// The maximum node-centric w-coordinate a horosphere can have such that the node in question
+    /// is still the owner of the horosphere.
+    // See `test_max_owned_horosphere_w()` for how this is computed.
+    const MAX_OWNED_HOROSPHERE_W: f32 = 5.9047837;
+}
+
+/// Represents a chunks's reference to a particular horosphere.
+pub struct HorosphereChunk {
+    /// The vector representing the horosphere in the perspective of the relevant chunk.
+    /// See `HorosphereNode::pos` for details.
+    pub pos: MVector<f32>,
+}
+
+impl HorosphereChunk {
+    /// Creates a `HorosphereChunk` based on a `HorosphereNode`
+    pub fn new(horosphere_node: &HorosphereNode, vertex: Vertex) -> Self {
+        HorosphereChunk {
+            pos: vertex.node_to_dual() * horosphere_node.pos,
+        }
+    }
+
+    /// Rasterizes the horosphere chunk into the given `VoxelData`
+    pub fn generate(&self, voxels: &mut VoxelData, chunk_size: u8) {
+        for x in 0..chunk_size {
+            for y in 0..chunk_size {
+                for z in 0..chunk_size {
+                    let pos = MVector::new(
```

Comment:
This calculation could probably be a function in `ChunkLayout`. I hope to try to do that in a PR that also tries to rename some terms to be more intuitive and switches the matrix transformations `x_to_y` to `y_from_x`.